pypsa/networkclustering.py

```diff
@@ -513,6 +514,85 @@ def rectangular_grid_clustering(network, divisions):
     busmap = busmap_by_rectangular_grid(network.buses, divisions)
     return get_clustering_from_busmap(network, busmap)
 
+################
+# Hierarchical Clustering
+def busmap_by_hac(network, n_clusters, buses_i=None, branch_components=["Line", "Link"], feature=None):
+    """
+    Create a busmap accroding to Hierarchical Agglomerative Clustering.
+
+    Parameters
+    ----------
+    network : pypsa.Network
+    n_clusters : int
+        Final number of clusters desired.
+    buses_i: None|pandas.Index
+        Subset of buses to cluster. If None, all buses are considered.
+    branch_components: List
+        Subset of all branch_components in the network. Defaults to ["Line", "Link"].
+    feature: None | pandas.DataFrame
+        Feature to be considered for the clustering.
+        The DataFrame must be indexed with buses_i.
+        If None, all buses have the same similarity.
+
+    Returns
+    -------
+    busmap : pandas.Series
+        Mapping of network.buses to k-means clusters (indexed by
+        non-negative integers).
+    """
+
+    if find_spec('sklearn') is None:
+        raise ModuleNotFoundError("Optional dependency 'sklearn' not found."
+            "Install via 'conda install -c conda-forge scikit-learn' "
+            "or 'pip install scikit-learn'")
+
+    from sklearn.cluster import AgglomerativeClustering as HAC
+
+    if buses_i is None:
+        buses_i = network.buses.index
+
+    if feature is None:
+        logger.warning(
+            "No feature is specified for Hierarchical Clustering. "
+            "Falling back to default, where all buses have equal similarity. "
+            "You can specify a feature as pandas.DataFrame indexed with buses_i."
+        )
+
+        feature = pd.DataFrame(index=buses_i, columns=[""], data=0)
+
+    buses_x = network.buses.index.get_indexer(buses_i)
+
+    A = network.adjacency_matrix(branch_components=branch_components).todense()
+    A = sp.sparse.coo_matrix(A[buses_x].T[buses_x].T)
+
+    labels = HAC(n_clusters=n_clusters,
+                 connectivity=A,
+                 affinity='euclidean',
+                 linkage='ward').fit_predict(feature)
```

Comment:
I would advocate for promoting `affinity` and `linkage` to kwargs.

To catch incompatible combinations, you can do something like:

```py
if linkage == 'ward':
    assert affinity == 'euclidian', "For Ward's method affinitiy needs to be euclidian."
```

If you do this, you could also add a selection of values that can be selected for affinity and linkage.

Reply:
ok, thanks for the feedback! I think I also prefer doing it that way; but if'll leave out the assertion error. this is already covered in hac itself :)

Reply:
ok, sounds good!